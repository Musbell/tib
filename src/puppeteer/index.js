import ChromeDetector from '../utils/detectors/chrome'
import Browser from '../browser'
import BrowserError from '../utils/error'
import Webpage from './webpage'

export default class PuppeteerBrowser extends Browser {
  async _loadDependencies() {
    super._loadDependencies()

    if (!PuppeteerBrowser.core) {
      PuppeteerBrowser.core = await this.loadDependency('puppeteer-core')
    }
  }

  setLogLevel(types) {
    const typeMap = {
      warning: 'warn'
    }

    if (types && typeof types === 'string') {
      types = [types]
    }

    this.hook('page:created', (page) => {
      /* eslint-disable no-console */
      page.on('console', (msg) => {
        const msgType = msg.type()
        let type = typeMap[msgType] || msgType
        if (!types || types.includes(msgType) || types.includes(type)) {
          if (!console[type]) {
            console.warn(`Unknown console type ${type}`)
            type = 'log'
          }
          console[type](msg.text())
        }
      })
      /* eslint-enable no-console */
    })
  }

  async _start(capabilities, ...args) {
    let executablePath = process.env.PUPPETEER_EXECUTABLE_PATH
    if (!executablePath) {
      const detector = new ChromeDetector()
      executablePath = detector.detect()
    }

    if (!executablePath) {
      throw new BrowserError(this, `Could not find a chrome executable`)
    }

    // https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteerlaunchoptions
    const launchOptions = {
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        ...args
      ],
      executablePath,
      ...capabilities
    }

    this.driver = await PuppeteerBrowser.core.launch(launchOptions)
  }

  async _close() {
    if (!this.driver) {
      return
    }

    await this.driver.close()
  }

  _page(url, readyCondition) {
    const page = new Webpage(this)
    return page.open(url, readyCondition)
  }
}
