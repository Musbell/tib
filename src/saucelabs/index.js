import SeleniumBrowser from '../selenium'
import SeleniumLogging from '../selenium/logging'
import BrowserError from '../utils/error'

export default class SauceLabsBrowser extends SeleniumLogging(SeleniumBrowser) {
  constructor(config) {
    config.xvfb = false

    super(config)

    this.hook('selenium:build:before', (builder) => {
      const user = this.getConfigProperty('user')
      const key = this.getConfigProperty('key')
      builder.usingServer(`http://${user}:${key}@ondemand.saucelabs.com:80/wd/hub`)
    })

    throw new BrowserError(this, `SauceLabs not yet implemented`)
  }

  getConfigProperty(property, capabilities = {}, defaulValue) {
    const envKey = `BROWSERSTACK_${property.toUpperCase()}`
    if (process.env[envKey]) {
      return process.env[envKey]
    }

    const confKey = `browserstack.${property}`

    if (capabilities[confKey]) {
      return capabilities[confKey]
    }

    if (this.config.SauceLabsLocal && this.config.SauceLabsLocal[property]) {
      return this.config.SauceLabsLocal[property]
    }

    if (this.config.SauceLabs && this.config.SauceLabs[property]) {
      return this.config.SauceLabs[property]
    }

    if (typeof defaulValue !== 'undefined') {
      return defaulValue
    }

    throw new Error(`${this.constructor.name} could not resolve required config property '${property}'`)
  }

  async _start(capabilities = {}) {
    this.addCapabilities({
      'browserstack.user': this.getConfigProperty('user', capabilities),
      'browserstack.key': this.getConfigProperty('key', capabilities),
      'browserstack.debug': this.getConfigProperty('debug', capabilities, false)
    })

    await super._start(capabilities)
  }

  setOS(name, version = '') {
    return this.addCapability('platform', `${name}${version ? ' ' : ''}${version || ''}`)
  }

  setDevice(deviceName) {
    return this.addCapability('device', deviceName)
  }
}
