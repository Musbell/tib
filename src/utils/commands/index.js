import Xvfb from './xvfb'
import BrowserStackLocal from './browserstack-local'

export {
  Xvfb,
  BrowserStackLocal
}
