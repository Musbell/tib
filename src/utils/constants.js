
// https://www.browserstack.com/automate/node#configure-capabilities
// https://wiki.saucelabs.com/display/DOCS/Platform+Configurator

export const browsers = [
  'chrome',
  'firefox',
  'ie',
  'edge',
  'safari'
]

export const drivers = [
  'puppeteer',
  'selenium'
]

export const providers = [
  'browserstack',
  'saucelabs'
]

export const providerOptions = {
  'local': ['browserstack']
}

export const os = [
  'windows',
  'linux',
  'mac',
  'macos',
  'macosx',
  'osx',
  'android',
  'ios'
]

export const osVersions = {
  windows: [
    'xp',
    '7',
    '8',
    '8.1',
    '10'
  ],
  apple: [
    '10.10',
    '10.11',
    '10.12',
    '10.13',
    '10.14',
    'snowleopard',
    'lion',
    'mountainlion',
    'mavericks',
    'yosemite',
    'elcapitan',
    'sierra',
    'highsierra',
    'mojave'
  ]
}
