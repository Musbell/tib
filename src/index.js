import Browser from './browser'
import * as commands from './utils/commands'

export async function browser(str, conf, autoStart = true) {
  const instance = Browser.get(str, conf)
  if (!autoStart) {
    return instance
  }

  return (await instance).start()
}

export {
  commands
}
