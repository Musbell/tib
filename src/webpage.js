import BrowserError from './utils/error'
import { abstractGuard, parseFunction, getDefaultHtmlCompiler } from './utils'

export default class Webpage {
  constructor(browser) {
    abstractGuard('Webpage', new.target)

    if (!browser || !browser.driver) {
      throw new BrowserError(this, 'Browser driver is required, has the browser been started succesfully?')
    }

    this.browser = browser
    this.driver = this.browser.driver
    this.userExtended = {}
  }

  returnProxy() {
    return new Proxy(this, {
      get(target, property) {
        target.browser.callHook('webpage:property', property)

        if (target.userExtended && target.userExtended[property]) {
          return target.userExtended[property]
        }

        if (target[property]) {
          return target[property]
        }

        if (target.page && target.page[property]) {
          return target.page[property]
        }

        return target.driver[property]
      }
    })
  }

  extend(extendWith = {}) {
    this.userExtended = extendWith
  }

  getHtmlCompiler() {
    if (this._htmlCompiler) {
      return this._htmlCompiler
    }

    let htmlCompiler
    if (typeof this.driver.htmlCompiler === 'function') {
      htmlCompiler = this.driver.htmlCompiler
    }

    if (!htmlCompiler || typeof htmlCompiler !== 'function') {
      htmlCompiler = getDefaultHtmlCompiler()
    }

    this._htmlCompiler = htmlCompiler
    return this._htmlCompiler
  }

  getBabelPresetOptions() {
    const presetOptions = {}

    const browser = this.browser.getBrowser()
    const version = this.browser.getBrowserVersion()

    if (browser && version) {
      presetOptions.targets = {
        [browser]: version
      }
    }

    return presetOptions
  }

  getHtml() {}

  runScript(...args) {}

  runAsyncScript(...args) {
    return this.runScript(...args)
  }

  async getElement(selector) {
    const html = await this.getElementHtml(selector)
    return this.getHtmlCompiler()(html)
  }

  async getElements(selector) {
    const htmls = await this.getElementsHtml(selector)
    const htmlCompiler = this.getHtmlCompiler()
    return htmls.map(html => htmlCompiler(html))
  }

  getElementFromPage(pageFunction, selector, ...args) {
    const parsedFn = parseFunction(pageFunction, this.getBabelPresetOptions())

    return this.runScript(
      /* istanbul ignore next */
      function (selector, fn, args) {
        return (new (Function.bind.apply(Function, fn))()).apply(null, [document.querySelector(selector)].concat(args))
      },
      selector,
      [null, ...parsedFn.args, parsedFn.body],
      args
    )
  }

  getElementsFromPage(pageFunction, selector, ...args) {
    const parsedFn = parseFunction(pageFunction, this.getBabelPresetOptions())

    return this.runScript(
      /* istanbul ignore next */
      function (selector, fn, args) {
        return (new (Function.bind.apply(Function, fn))()).apply(null, [Array.prototype.slice.call(document.querySelectorAll(selector))].concat(args))
      },
      selector,
      [null, ...parsedFn.args, parsedFn.body],
      args
    )
  }

  getElementCount(selector) {
    /* istanbul ignore next */
    const pageFn = els => els.length
    return this.getElementsFromPage(pageFn, selector)
  }

  getElementHtml(selector) {
    /* istanbul ignore next */
    const pageFn = el => el.outerHTML
    return this.getElementFromPage(pageFn, selector)
  }

  getElementsHtml(selector) {
    /* istanbul ignore next */
    const pageFn = els => els.map(el => el.outerHTML)
    return this.getElementsFromPage(pageFn, selector)
  }

  getAttribute(selector, attribute) {
    /* istanbul ignore next */
    const pageFn = (el, attribute) => el.getAttribute(attribute)
    return this.getElementFromPage(pageFn, selector, attribute)
  }

  getAttributes(selector, attribute) {
    /* istanbul ignore next */
    const pageFn = (els, attribute) => els.map(el => el.getAttribute(attribute))
    return this.getElementsFromPage(pageFn, selector, attribute)
  }

  getText(selector) {
    /* istanbul ignore next */
    const pageFn = el => el.textContent
    return this.getElementFromPage(pageFn, selector)
  }

  getTexts(selector) {
    /* istanbul ignore next */
    const pageFn = els => els.map(el => el.textContent)
    return this.getElementsFromPage(pageFn, selector)
  }

  clickElement(selector) {
    /* istanbul ignore next */
    const pageFn = el => el.click()
    return this.getElementFromPage(pageFn, selector)
  }
}
