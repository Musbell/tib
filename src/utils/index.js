import parseFn from 'parse-function'
import { compile } from 'vue-template-compiler'
import BrowserError from './error'
import * as constants from './constants'

export function abstractGuard(className, { name } = {}) {
  if (className === name) {
    throw new BrowserError(`Do not use abstract class '${className}' directly`)
  }
}

// TODO: add more test framework checks like sinon?
export function enableTimers() {
  // find Jest fingerprint
  if (process.env.JEST_WORKER_ID) {
    try {
      // call useFakeTimers first as it seems there is no way to determine
      // if fake timers are used at all and otherwise runOnlyPendingTimers
      // will fail with a warning from within Jest
      // -> Disabled because we probably dont care about pending timers
      // jest.useFakeTimers()
      // jest.runOnlyPendingTimers()
      jest.useRealTimers()
    } catch (e) {
      /* istanbul ignore next */
      throw new BrowserError(`Enabling real timers failed: ${e.message}`)
    }
  }
}

/* Not used at the moment
export function findNodeByType(node, type) {
  if (Array.isArray(node.body)) {
    for (const k in node.body) {
      if (findNodeByType(node.body[k], type)) {
        return true
      }
    }
  } else if (node.body) {
    return findNodeByType(node.body, type)
  } else if (node.type === type) {
    return true
  }
  return false
}
/**/

let fnParser
export function parseFunction(fn, needsReturn) {
  if (typeof fn !== 'function') {
    throw new BrowserError(`parseFunction expects the first argument to be a function, received '${typeof fn}' instead`)
  }

  if (!fnParser) {
    fnParser = parseFn()
    fnParser.use((self) => {
      return (node, result) => {
        if (fnParser.needsReturn && node.type === 'ArrowFunctionExpression' && node.body.type !== 'ReturnStatement' && node.body.type !== 'BlockStatement') {
          self.define(result, 'body', `return ${result.body}`)
        }
      }
    })
  }

  fnParser.needsReturn = needsReturn

  return fnParser.parse(fn)
}

export function getDefaultHtmlCompiler() {
  return html => compile(html).ast
}

export async function loadDependency(dependency) {
  try {
    return await import(dependency).then(m => m.default || m)
  } catch (e) {
    throw new BrowserError(`Could not import the required dependency '${dependency}'
(error: ${e.message})

Please install the dependency with:
$ npm install ${dependency}
or
$ yarn add ${dependency}
    `)
  }
}

export function getBrowserConfigFromString(browserString) {
  const browserConfig = {}

  browserString
    .split('/')
    .filter(s => !!s)
    .map(s => s.toLowerCase()
      // replace first occurence with =
      // -> Chrome 39 => chrome=39
      // -> Mac OS X=High Sierra => Mac OS X=High Sierra
      .replace(/([\s:])/, (m, $1, o, s) => (s.includes('=') ? $1 : '='))
    )
    .forEach((s) => {
      let [ key, value ] = s.split('=', 2) // eslint-disable-line prefer-const
      key = key.replace(/\s+/, '')

      if (constants.browsers.includes(key)) {
        browserConfig.browser = key

        if (value) {
          browserConfig.browserVersion = value.replace(/\s+/g, '')
        }
        return
      }

      if (constants.browserOptions.includes(key)) {
        browserConfig[key] = typeof value === 'undefined' ? true : value
        return
      }

      if (constants.os.includes(key)) {
        browserConfig.os = key

        if (value) {
          browserConfig.osVersion = value
        }
        return
      }

      if (constants.drivers.includes(key)) {
        browserConfig.driver = key
        return
      }

      if (constants.providers.includes(key)) {
        browserConfig.provider = key
        return
      }

      if (constants.browserVariants[key]) {
        browserConfig.browserVariant = key
        return
      }

      if (key === 'device') {
        browserConfig.device = value
        return
      }

      if (!value) {
        const screenResolution = key.split(/(\d+)x(\d+)/i)
        if (screenResolution && screenResolution.length > 1) {
          browserConfig.window = {
            width: parseInt(screenResolution[1]),
            height: parseInt(screenResolution[2])
          }
        }
      }
    })

  return browserConfig
}

export function getBrowserImportFromConfig(browserConfig) {
  const importPath = []

  if (browserConfig.browser) {
    importPath.push(browserConfig.browser)
  }

  if (browserConfig.driver && !browserConfig.provider) {
    importPath.push(browserConfig.driver)
  }

  if (browserConfig.provider) {
    importPath.push(browserConfig.provider)
  }

  if (browserConfig.browserVariant) {
    importPath.push(browserConfig.browserVariant)
  }

  return importPath.join('/')
}
