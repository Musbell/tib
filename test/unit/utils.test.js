import * as utils from '../../src/utils'

describe('utils', () => {
  test('browser strings', () => {
    const browserStrings = [
      'chrome',
      'chrome/windows',
      'chrome/windows 8.1',
      'chrome/windows=8',
      'chrome/windows:7',
      'macos high sierra',
      'mac os=high sierra',
      'firefox',
      'puppeteer/linux',
      'selenium',
      '/windows 7/chrome 39/browserstack/local/1280x1024',
      'android 8/device=Android Emulator'
    ]

    browserStrings.forEach((s) => {
      /*
       * !!! check the snapshot carefully before updating !!!
       */
      // console.log(utils.getBrowserConfigFromString(s))
      expect(utils.getBrowserConfigFromString(s)).toMatchSnapshot()
    })
  })
})
