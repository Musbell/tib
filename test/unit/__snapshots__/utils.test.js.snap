// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils browser strings 1`] = `
Object {
  "browser": "chrome",
}
`;

exports[`utils browser strings 2`] = `
Object {
  "browser": "chrome",
  "os": "windows",
}
`;

exports[`utils browser strings 3`] = `
Object {
  "browser": "chrome",
  "os": "windows",
  "osVersion": "8.1",
}
`;

exports[`utils browser strings 4`] = `
Object {
  "browser": "chrome",
  "os": "windows",
  "osVersion": "8",
}
`;

exports[`utils browser strings 5`] = `
Object {
  "browser": "chrome",
  "os": "windows",
  "osVersion": "7",
}
`;

exports[`utils browser strings 6`] = `
Object {
  "browser": "chrome",
  "os": "macos",
  "osVersion": "high sierra",
}
`;

exports[`utils browser strings 7`] = `
Object {
  "browser": "firefox",
  "os": "macos",
  "osVersion": "high sierra",
}
`;

exports[`utils browser strings 8`] = `
Object {
  "browser": "firefox",
  "headless": true,
}
`;

exports[`utils browser strings 9`] = `
Object {
  "driver": "puppeteer",
  "os": "linux",
}
`;

exports[`utils browser strings 10`] = `
Object {
  "driver": "selenium",
}
`;

exports[`utils browser strings 11`] = `
Object {
  "driver": "puppeteer",
}
`;

exports[`utils browser strings 12`] = `
Object {
  "browser": "chrome",
  "browserVariant": "local",
  "browserVersion": "39",
  "os": "windows",
  "osVersion": "7",
  "provider": "browserstack",
  "window": Object {
    "height": 1024,
    "width": 1280,
  },
}
`;

exports[`utils browser strings 13`] = `
Object {
  "device": "android emulator",
  "os": "android",
  "osVersion": "8",
  "provider": "browserstack",
}
`;

exports[`utils default html compiler should work 1`] = `
Object {
  "attrs": Array [
    Object {
      "dynamic": undefined,
      "name": "id",
      "value": "\\"test\\"",
    },
  ],
  "attrsList": Array [
    Object {
      "name": "id",
      "value": "test",
    },
  ],
  "attrsMap": Object {
    "id": "test",
  },
  "children": Array [],
  "parent": undefined,
  "plain": false,
  "rawAttrsMap": Object {},
  "static": true,
  "staticInFor": false,
  "staticRoot": false,
  "tag": "div",
  "type": 1,
}
`;
